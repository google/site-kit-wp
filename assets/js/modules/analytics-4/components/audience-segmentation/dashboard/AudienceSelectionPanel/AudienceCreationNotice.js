/**
 * Audience Selection Panel AudienceCreationNotice component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { useCallback, useEffect, useState } from '@wordpress/element';
import { addQueryArgs } from '@wordpress/url';

/**
 * Internal dependencies
 */
import { useDispatch, useInViewSelect, useSelect } from 'googlesitekit-data';
import useViewContext from '../../../../../../hooks/useViewContext';
import { trackEvent } from '../../../../../../util';
import {
	AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG,
	AUDIENCE_CREATION_FORM,
	AUDIENCE_CREATION_NOTICE_SLUG,
	AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG,
	AUDIENCE_SELECTION_PANEL_OPENED_KEY,
} from './constants';
import { CORE_FORMS } from '../../../../../../googlesitekit/datastore/forms/constants';
import { CORE_SITE } from '../../../../../../googlesitekit/datastore/site/constants';
import { CORE_USER } from '../../../../../../googlesitekit/datastore/user/constants';
import { CORE_UI } from '../../../../../../googlesitekit/datastore/ui/constants';
import {
	EDIT_SCOPE,
	MODULES_ANALYTICS_4,
	SITE_KIT_AUDIENCE_DEFINITIONS,
} from '../../../../datastore/constants';
import { ERROR_CODE_MISSING_REQUIRED_SCOPE } from '../../../../../../util/errors';
import Link from '../../../../../../components/Link';
import CloseIcon from '../../../../../../../svg/icons/close.svg';
import SpinnerButton, {
	SPINNER_POSITION,
} from '../../../../../../googlesitekit/components-gm2/SpinnerButton';
import SubtleNotification, {
	VARIANTS,
} from '../../../../../../components/notifications/SubtleNotification';
import AudienceCreationErrorNotice from './AudienceCreationErrorNotice';

export default function AudienceCreationNotice() {
	const viewContext = useViewContext();

	const [ isCreatingAudience, setIsCreatingAudience ] = useState( false );

	const siteKitConfigurableAudiences = useInViewSelect( ( select ) => {
		const { getConfigurableAudiences } = select( MODULES_ANALYTICS_4 );

		const audiences = getConfigurableAudiences();

		if ( undefined === audiences ) {
			return undefined;
		}

		if ( ! audiences.length ) {
			return [];
		}

		return audiences.filter(
			( { audienceType } ) => audienceType === 'SITE_KIT_AUDIENCE'
		);
	} );

	const { dismissItem } = useDispatch( CORE_USER );
	const { setValue } = useDispatch( CORE_UI );

	const isItemDismissed = useInViewSelect( ( select ) =>
		select( CORE_USER ).isItemDismissed( AUDIENCE_CREATION_NOTICE_SLUG )
	);
	const isEditScopeNoticeDismissed = useInViewSelect( ( select ) =>
		select( CORE_USER ).isItemDismissed(
			AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG
		)
	);
	const hasAnalytics4EditScope = useInViewSelect( ( select ) =>
		select( CORE_USER ).hasScope( EDIT_SCOPE )
	);
	const isOpen = useSelect( ( select ) =>
		select( CORE_UI ).getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
	);

	const onCloseClick = () => {
		dismissItem( AUDIENCE_CREATION_NOTICE_SLUG );
	};

	const redirectURL = addQueryArgs( global.location.href, {
		notification: 'audience_segmentation',
	} );

	const { setValues } = useDispatch( CORE_FORMS );
	const { setPermissionScopeError } = useDispatch( CORE_USER );
	const { createAudience, syncAvailableAudiences } =
		useDispatch( MODULES_ANALYTICS_4 );

	const isCreatingAudienceFromOAuth = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( AUDIENCE_CREATION_FORM, 'autoSubmit' )
	);

	const failedAudienceToCreate = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue(
			AUDIENCE_CREATION_FORM,
			'audienceToCreate'
		)
	);

	const [ apiErrors, setApiErrors ] = useState( [] );

	const handleCreateAudience = useCallback(
		async ( audienceSlug ) => {
			setIsCreatingAudience( audienceSlug );

			if ( ! hasAnalytics4EditScope ) {
				setValues( AUDIENCE_CREATION_FORM, {
					autoSubmit: true,
					audienceToCreate: audienceSlug,
				} );

				// Set permission scope error to trigger OAuth flow.
				setPermissionScopeError( {
					code: ERROR_CODE_MISSING_REQUIRED_SCOPE,
					message: __(
						'Additional permissions are required to create a new audience in Analytics.',
						'google-site-kit'
					),
					data: {
						status: 403,
						scopes: [ EDIT_SCOPE ],
						skipModal: true,
						redirectURL,
					},
				} );

				return;
			}

			setValues( AUDIENCE_CREATION_FORM, {
				autoSubmit: false,
				audienceToCreate: undefined,
			} );

			const { error } = await createAudience(
				SITE_KIT_AUDIENCE_DEFINITIONS[ audienceSlug ]
			);

			if ( !! error ) {
				setApiErrors( [ error ] );
			} else {
				setApiErrors( [] );
			}

			await syncAvailableAudiences();

			setIsCreatingAudience( false );

			if ( ! error ) {
				setValue( AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG, true );
			}
		},
		[
			hasAnalytics4EditScope,
			createAudience,
			syncAvailableAudiences,
			setValues,
			setPermissionScopeError,
			redirectURL,
			setValue,
		]
	);

	const handleDismissEditScopeNotice = () => {
		trackEvent(
			`${ viewContext }_audiences-sidebar-create-audiences`,
			'dismiss_oauth_notice'
		).finally( () => {
			dismissItem( AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG );
		} );
	};

	const setupErrorCode = useSelect( ( select ) =>
		select( CORE_SITE ).getSetupErrorCode()
	);

	const hasOAuthError =
		isCreatingAudienceFromOAuth && setupErrorCode === 'access_denied';

	useEffect( () => {
		async function createAudienceFromOAuth() {
			if ( hasAnalytics4EditScope && isCreatingAudienceFromOAuth ) {
				setValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY, true );
				await handleCreateAudience( failedAudienceToCreate );
			}
		}

		createAudienceFromOAuth();
	}, [
		failedAudienceToCreate,
		handleCreateAudience,
		hasAnalytics4EditScope,
		isCreatingAudienceFromOAuth,
		setValue,
	] );

	// Show the notice if the user has no site kit audiences, or has
	// created one, and the user has not dismissed it.
	const shouldShowNotice =
		! isItemDismissed && siteKitConfigurableAudiences?.length < 2;

	// Track an event when the notice is viewed.
	useEffect( () => {
		if ( isOpen && shouldShowNotice ) {
			trackEvent(
				`${ viewContext }_audiences-sidebar-create-audiences`,
				'view_notice'
			);
		}
	}, [ isOpen, shouldShowNotice, viewContext ] );

	// Track an event when the OAuth notice is viewed.
	useEffect( () => {
		if (
			isOpen &&
			! hasAnalytics4EditScope &&
			! isEditScopeNoticeDismissed
		) {
			trackEvent(
				`${ viewContext }_audiences-sidebar-create-audiences`,
				'view_oauth_notice'
			);
		}
	}, [
		hasAnalytics4EditScope,
		isEditScopeNoticeDismissed,
		isOpen,
		viewContext,
	] );

	if ( ! shouldShowNotice ) {
		return null;
	}

	const siteKitAvailableAudiences = Object.keys(
		SITE_KIT_AUDIENCE_DEFINITIONS
	).filter(
		( audienceSlug ) =>
			! siteKitConfigurableAudiences.some(
				( configuredAudience ) =>
					configuredAudience.audienceSlug === audienceSlug
			)
	);

	return (
		<div className="googlesitekit-audience-selection-panel__audience-creation-notice">
			<div className="googlesitekit-audience-selection-panel__audience-creation-notice-header">
				<p className="googlesitekit-audience-selection-panel__audience-creation-notice-title">
					{ __(
						'Create groups suggested by Site Kit',
						'google-site-kit'
					) }
				</p>

				<Link
					className="googlesitekit-audience-selection-panel__audience-creation-notice-close"
					onClick={ onCloseClick }
					linkButton
				>
					<CloseIcon width="15" height="15" />
				</Link>
			</div>
			<div className="googlesitekit-audience-selection-panel__audience-creation-notice-body">
				{ siteKitAvailableAudiences &&
					siteKitAvailableAudiences.map( ( audienceSlug ) => (
						<div
							key={ audienceSlug }
							className="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
						>
							<div className="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details">
								<h3>
									{
										SITE_KIT_AUDIENCE_DEFINITIONS[
											audienceSlug
										].displayName
									}
								</h3>
								<p className="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description">
									{
										SITE_KIT_AUDIENCE_DEFINITIONS[
											audienceSlug
										].description
									}
								</p>
							</div>
							<div className="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button">
								<SpinnerButton
									spinnerPosition={ SPINNER_POSITION.BEFORE }
									onClick={ () => {
										trackEvent(
											`${ viewContext }_audiences-sidebar-create-audiences`,
											'create_audience',
											audienceSlug
										).finally( () => {
											handleCreateAudience(
												audienceSlug
											);
										} );
									} }
									isSaving={
										isCreatingAudience === audienceSlug
									}
								>
									{ __( 'Create', 'google-site-kit' ) }
								</SpinnerButton>
							</div>
						</div>
					) ) }
			</div>
			{ ! hasAnalytics4EditScope && ! isEditScopeNoticeDismissed && (
				<div className="googlesitekit-audience-selection-panel__audience-creation-notice-info">
					<SubtleNotification
						title={ __(
							'Creating these groups require more data tracking. You will be directed to update your Analytics property.',
							'google-site-kit'
						) }
						dismissLabel={ __( 'Got it', 'google-site-kit' ) }
						onDismiss={ handleDismissEditScopeNotice }
						variant={ VARIANTS.WARNING }
						hideIcon
					/>
				</div>
			) }
			{ ( apiErrors.length > 0 || hasOAuthError ) && (
				<AudienceCreationErrorNotice
					apiErrors={ apiErrors }
					hasOAuthError={ hasOAuthError }
				/>
			) }
		</div>
	);
}
