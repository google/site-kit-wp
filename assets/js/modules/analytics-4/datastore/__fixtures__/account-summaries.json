{
  "accountSummaries": [
    {
      "account": "accounts/123456789",
      "displayName": "Example Org",
      "name": "accountSummaries/123456789",
      "_id": "123456789",
      "propertySummaries": []
    },
    {
      "account": "accounts/987654321",
      "displayName": "Example Com",
      "name": "accountSummaries/987654321",
      "propertySummaries": [
        {
          "displayName": "Example Property",
          "parent": "accounts/987654321",
          "property": "properties/1122334455",
          "_id": "1122334455",
          "_accountID": "987654321"
        }
      ],
      "_id": "987654321"
    },
    {
      "account": "accounts/152925174",
      "displayName": "Example Net",
      "name": "accountSummaries/152925174",
      "propertySummaries": [
        {
          "displayName": "Example Property",
          "parent": "accounts/152925174",
          "property": "properties/1122334455",
          "_id": "1122334455",
          "_accountID": "152925174"
        }
      ],
      "_id": "152925174"
    }
  ],
  "nextPageToken": null
}
