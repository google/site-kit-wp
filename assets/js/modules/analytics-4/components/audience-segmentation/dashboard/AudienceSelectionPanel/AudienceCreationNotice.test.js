/**
 * Audience Selection Panel AudienceCreationNotice tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	render,
	createTestRegistry,
	provideModules,
	provideModuleRegistrations,
	provideUserInfo,
	provideUserAuthentication,
	fireEvent,
	muteFetch,
	act,
} from '../../../../../../../../tests/js/test-utils';
import { availableAudiences } from '../../../../datastore/__fixtures__';
import * as tracking from '../../../../../../util/tracking';
import {
	AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG,
	AUDIENCE_CREATION_NOTICE_SLUG,
	AUDIENCE_SELECTION_PANEL_OPENED_KEY,
} from './constants';
import { CORE_UI } from '../../../../../../googlesitekit/datastore/ui/constants';
import { CORE_USER } from '../../../../../../googlesitekit/datastore/user/constants';
import {
	EDIT_SCOPE,
	MODULES_ANALYTICS_4,
} from '../../../../datastore/constants';
import { VIEW_CONTEXT_MAIN_DASHBOARD } from '../../../../../../googlesitekit/constants';
import AudienceCreationNotice from './AudienceCreationNotice';

const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
mockTrackEvent.mockImplementation( () => Promise.resolve() );

describe( 'AudienceCreationNotice', () => {
	let registry;

	const createAudienceEndpoint = new RegExp(
		'^/google-site-kit/v1/modules/analytics-4/data/create-audience'
	);
	const syncAvailableAudiencesEndpoint = new RegExp(
		'^/google-site-kit/v1/modules/analytics-4/data/sync-audiences'
	);
	const dismissItemEndpoint = new RegExp(
		'^/google-site-kit/v1/core/user/data/dismiss-item'
	);

	beforeEach( () => {
		registry = createTestRegistry();
		provideUserInfo( registry );
		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
		] );
		provideModuleRegistrations( registry );

		registry
			.dispatch( CORE_UI )
			.setValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY, true );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );

		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );
	} );

	afterEach( () => {
		mockTrackEvent.mockClear();
	} );

	it( 'should render null if no audiences are available', () => {
		const { container } = render( <AudienceCreationNotice />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should render null if the user has dismissed the notice', async () => {
		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedItems( [ AUDIENCE_CREATION_NOTICE_SLUG ] );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: availableAudiences.reduce( ( acc, { name } ) => {
					acc[ name ] = 20201220;

					return acc;
				}, {} ),
				customDimension: {},
				property: {},
			} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableAudiences,
		} );

		const { container, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should render the notice if the user has not dismissed the notice and there are 2 available audiences', async () => {
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: availableAudiences.reduce( ( acc, { name } ) => {
					acc[ name ] = 20201220;

					return acc;
				}, {} ),
				customDimension: {},
				property: {},
			} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableAudiences: [],
		} );

		const { container, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await waitForRegistry();

		// Verify the edit scope notice is not displayed.
		expect( container ).not.toHaveTextContent(
			'Creating these groups require more data tracking. You will be directed to update your Analytics property.'
		);

		expect( container ).toMatchSnapshot();

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			`${ VIEW_CONTEXT_MAIN_DASHBOARD }_audiences-sidebar-create-audiences`,
			'view_notice'
		);
	} );

	it( 'should render the notice if the user has not dismissed the notice and there is 1 available audience', async () => {
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: availableAudiences.reduce( ( acc, { name } ) => {
					acc[ name ] = 20201220;

					return acc;
				}, {} ),
				customDimension: {},
				property: {},
			} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableAudiences: availableAudiences.filter(
				( { displayName } ) => displayName !== 'Returning visitors'
			),
		} );

		const { container, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await waitForRegistry();

		// Verify the edit scope notice is not displayed.
		expect( container ).not.toHaveTextContent(
			'Creating these groups require more data tracking. You will be directed to update your Analytics property.'
		);

		expect( container ).toMatchSnapshot();

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			`${ VIEW_CONTEXT_MAIN_DASHBOARD }_audiences-sidebar-create-audiences`,
			'view_notice'
		);
	} );

	it.each( [
		[ 'new-visitors', 'New visitors' ],
		[ 'returning-visitors', 'Returning visitors' ],
	] )(
		'should track an event with label %s when creating the "%s" audience',
		async ( label, audienceDisplayName ) => {
			const filteredAudiences = availableAudiences.filter(
				( { displayName } ) => displayName !== audienceDisplayName
			);

			fetchMock.post( syncAvailableAudiencesEndpoint, {
				status: 200,
				body: filteredAudiences,
			} );

			muteFetch( createAudienceEndpoint );

			registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
				configuredAudiences: [],
				isAudienceSegmentationWidgetHidden: false,
				didSetAudiences: true,
			} );

			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveResourceDataAvailabilityDates( {
					audience: availableAudiences.reduce( ( acc, { name } ) => {
						acc[ name ] = 20201220;

						return acc;
					}, {} ),
					customDimension: {},
					property: {},
				} );

			registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
				accountID: '12345',
				propertyID: '34567',
				measurementID: '56789',
				webDataStreamID: '78901',
				availableAudiences: filteredAudiences,
			} );

			const { getByRole, waitForRegistry } = render(
				<AudienceCreationNotice />,
				{
					registry,
					viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
				}
			);

			await waitForRegistry();

			const button = getByRole( 'button', { name: /create/i } );

			// eslint-disable-next-line require-await
			await act( async () => {
				fireEvent.click( button );
			} );

			expect( mockTrackEvent ).toHaveBeenCalledWith(
				`${ VIEW_CONTEXT_MAIN_DASHBOARD }_audiences-sidebar-create-audiences`,
				'create_audience',
				label
			);
		}
	);

	it( 'should render the missing scope notice if the user does not have the edit scope', async () => {
		provideUserAuthentication( registry, {
			grantedScopes: [],
		} );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: availableAudiences.reduce( ( acc, { name } ) => {
					acc[ name ] = 20201220;

					return acc;
				}, {} ),
				customDimension: {},
				property: {},
			} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableAudiences: [],
		} );

		const { container, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await waitForRegistry();

		// Verify the edit scope notice is displayed.
		expect( container ).toHaveTextContent(
			'Creating these groups require more data tracking. You will be directed to update your Analytics property.'
		);

		expect( container ).toMatchSnapshot();

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			`${ VIEW_CONTEXT_MAIN_DASHBOARD }_audiences-sidebar-create-audiences`,
			'view_oauth_notice'
		);
	} );

	it( 'should track an event when the missing scope notice is dismissed', async () => {
		muteFetch( dismissItemEndpoint );

		provideUserAuthentication( registry, {
			grantedScopes: [],
		} );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: availableAudiences.reduce( ( acc, { name } ) => {
					acc[ name ] = 20201220;

					return acc;
				}, {} ),
				customDimension: {},
				property: {},
			} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableAudiences: [],
		} );

		const { getByRole, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await waitForRegistry();

		const button = getByRole( 'button', { name: /Got it/i } );

		fireEvent.click( button );

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			`${ VIEW_CONTEXT_MAIN_DASHBOARD }_audiences-sidebar-create-audiences`,
			'dismiss_oauth_notice'
		);
	} );

	it( 'should not render the missing scope notice if the user does not have the edit scope but the notice has been dismissed', async () => {
		provideUserAuthentication( registry, {
			grantedScopes: [],
		} );
		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedItems( [
				AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG,
			] );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: availableAudiences.reduce( ( acc, { name } ) => {
					acc[ name ] = 20201220;

					return acc;
				}, {} ),
				customDimension: {},
				property: {},
			} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableAudiences: [],
		} );

		const { container, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
			}
		);

		await waitForRegistry();

		// Verify the edit scope notice is not displayed.
		expect( container ).not.toHaveTextContent(
			'Creating these groups require more data tracking. You will be directed to update your Analytics property.'
		);

		expect( container ).toMatchSnapshot();
	} );
} );
