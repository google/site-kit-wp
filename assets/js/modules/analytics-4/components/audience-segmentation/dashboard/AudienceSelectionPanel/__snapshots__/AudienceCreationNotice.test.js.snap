// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudienceCreationNotice should not render the missing scope notice if the user does not have the edit scope but the notice has been dismissed 1`] = `
<div>
  <div
    class="googlesitekit-audience-selection-panel__audience-creation-notice"
  >
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-header"
    >
      <p
        class="googlesitekit-audience-selection-panel__audience-creation-notice-title"
      >
        Create groups suggested by Site Kit
      </p>
      <button
        class="googlesitekit-cta-link googlesitekit-audience-selection-panel__audience-creation-notice-close googlesitekit-cta-link--link-button"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          <svg />
        </span>
      </button>
    </div>
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-body"
    >
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            New visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who visited the site for the first time
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            Returning visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who have visited your site at least once before
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AudienceCreationNotice should render the missing scope notice if the user does not have the edit scope 1`] = `
<div>
  <div
    class="googlesitekit-audience-selection-panel__audience-creation-notice"
  >
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-header"
    >
      <p
        class="googlesitekit-audience-selection-panel__audience-creation-notice-title"
      >
        Create groups suggested by Site Kit
      </p>
      <button
        class="googlesitekit-cta-link googlesitekit-audience-selection-panel__audience-creation-notice-close googlesitekit-cta-link--link-button"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          <svg />
        </span>
      </button>
    </div>
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-body"
    >
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            New visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who visited the site for the first time
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            Returning visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who have visited your site at least once before
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-info"
    >
      <div
        class="googlesitekit-subtle-notification googlesitekit-subtle-notification--warning"
      >
        <div
          class="googlesitekit-subtle-notification__content"
        >
          <p>
            Creating these groups require more data tracking. You will be directed to update your Analytics property.
          </p>
        </div>
        <div
          class="googlesitekit-subtle-notification__action"
        >
          <button
            class="mdc-button mdc-button--tertiary"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Got it
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AudienceCreationNotice should render the notice if the user has not dismissed the notice and there are 2 available audiences 1`] = `
<div>
  <div
    class="googlesitekit-audience-selection-panel__audience-creation-notice"
  >
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-header"
    >
      <p
        class="googlesitekit-audience-selection-panel__audience-creation-notice-title"
      >
        Create groups suggested by Site Kit
      </p>
      <button
        class="googlesitekit-cta-link googlesitekit-audience-selection-panel__audience-creation-notice-close googlesitekit-cta-link--link-button"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          <svg />
        </span>
      </button>
    </div>
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-body"
    >
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            New visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who visited the site for the first time
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            Returning visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who have visited your site at least once before
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AudienceCreationNotice should render the notice if the user has not dismissed the notice and there is 1 available audience 1`] = `
<div>
  <div
    class="googlesitekit-audience-selection-panel__audience-creation-notice"
  >
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-header"
    >
      <p
        class="googlesitekit-audience-selection-panel__audience-creation-notice-title"
      >
        Create groups suggested by Site Kit
      </p>
      <button
        class="googlesitekit-cta-link googlesitekit-audience-selection-panel__audience-creation-notice-close googlesitekit-cta-link--link-button"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          <svg />
        </span>
      </button>
    </div>
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-body"
    >
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            Returning visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who have visited your site at least once before
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
